_posts/2020/2020-01-10-gasper-part-1.md

```diff
@@ -0,0 +1,89 @@
+* Kaen 🔥(the worker service for creating/managing databases)
+* Enrai ⚡️(the reverse proxy server)
+* Hikari 💡(the service for internal DNS resolution)
+* Iwa 🗿(the service that handles ssh into the containers)
+
+Let’s start from the very core itself: How to package? Multiple apps in a single node or one app per node. The former just sends security on a walk, and the latter one brings about a lot of hardware overheads. 
+
+#### H're cometh Docker...
+
+Docker is a service that deploys containers. These containers are a sort of OS-level virtualization. Docker comes with incredible configuration options from the allocation of CPU and Memory resources to networking inside a container. It proved a perfect solution to our problem; containers having a low resource overhead, with isolation for configuration files and dependencies.
+
+<div style="padding-left:10%;" markdown="1">
+  ![Docker](/images/posts/gasper-part-1/docker.jpg)
+</div>
+
+With docker containerization at its core, apps now can be packaged and built. But we wanted to get rid of that nasty Dockerfile and docker-compose, right? Let’s automate things a bit.
+ 
+With all the mentioned components in place, the pipeline for deployment follows:
+1. Download the application source code
+2. Mount this source code as a volume for the container
+3. Bind the ports of the host and the container
+4. Start the docker container with this configuration
+5. Install the dependencies
+6. Fire up the application 
+
+And hola, your application is live!
+
+Golang’s Docker Engine library was enough to handle steps 2 through 6. But here comes an issue, the whole process took a whopping 15s, even for simple template apps. We needed to optimize that.
+
+#### Go, go, go… making thy w'rk easy
+
+Go, besides the fact that it came from Google, is famous for its efficient realization of concurrency using Goroutines. And there it was, we took the two slowest processes, steps 1 and 4, and put them in goroutines. The deployment times came down from ~15s to ~5.5s. Further optimizations pulled it down to 1.5s. Phew! That was fast.
+
+All of the above steps are handled by the service Mizu. Databases are managed separately by the service, Kaen. More details on them in the next blog.
+
+## The final take:
+
+The next step was the reverse proxying of requests. For that, we have Enrai, the reverse proxy server. As a matter of fact, only Enrai is given a public IP and the rest of the apps are proxied through it.\\
+Gasper boasts an internal DNS resolution service, too, Hikari. It allots domain names to the apps and databases and maintains the DNS records.\\
+For those who prefer a hands-on deployment experience, Gasper provides SSH into the containers along with a virtual terminal for ease. Iwa plays its role here.
+
+This marks the end of the deployment journey. The user is always able to restart/rebuild the app in case of failure. Still, he can always use SSH and get his hands dirty 💪.
```

Comment:
```suggestion
This marks the end of the deployment journey. The user is always able to restart/rebuild the app in case of failure, and there's always SSH if they want to get their hands dirty 💪.
```